pacstall/api/config_facade.py

```diff
@@ -0,0 +1,259 @@
+        )
+        return False
+    if type(conf[attribute_name]) is not ttype:
+        fancy(
+            "error",
+            f"Repository '{repo_name}' attribute '{attribute_name}' must be of type '{ttype}'",
+        )
+        return False
+    return True
+
+
+def __raise_unreachable() -> NoReturn:
+    """
+    Small hack to help the type checker in complex cases.
+    """
+    raise Exception("Unreachable code. This will never be raised.")
+
+
+def read_config() -> Tuple[Optional[ErrorCodes], Optional[List[RepositoryConfig]]]:
+    """
+    Reads and parses the repository list.
+
+    Returns
+    -------
+    A tuple consisting of the error code, and the repository list.
+    If the error code is `None` then the repository list will *not* be `None` and vice-versa.
+    """
+    try:
+        config_dict: Optional[Dict[str, Dict[str, Dict[str, Optional[str]]]]] = None
+        try:
+            with open(PACSTALL_CONFIG_PATH) as file:
+                config_dict = tomli.load(file)  # type: ignore[arg-type]
+        except OSError as error:
+            fancy(
+                "error",
+                f"Could not read repositories from file '{PACSTALL_CONFIG_PATH}'.\n{error}",
+            )
+
+            return (ErrorCodes.NO_INPUT_ERROR, None)
+
+        parsed_repo_list: List[RepositoryConfig] = []
+        for (repo_name, conf) in config_dict["repository"].items():
+            if not __validate_attribute(
+                repo_name, conf, attribute_name="url", ttype=str
+            ) or not __validate_attribute(
+                repo_name, conf, attribute_name="branch", ttype=str
+            ):
+                return (ErrorCodes.CONFIG_ERROR, None)
+
+            # Help type checker recognize `[conf["url"]` is not None
+            url = conf["url"] if conf["url"] is not None else __raise_unreachable()
+            parsed_url: Optional[str] = parse_url(url)
+            if parsed_url is None:
+                fancy(
+                    "error",
+                    f"Repository '{repo_name}' has invalid attribute 'url': {url}",
+                )
+                return (ErrorCodes.CONFIG_ERROR, None)
+
+            # Help type checker recognize `[conf["branch"]` is not None
+            branch = (
+                conf["branch"] if conf["branch"] is not None else __raise_unreachable()
+            )
+            repo_entry = RepositoryConfig(repo_name, parsed_url, branch, url)
+
+            if not is_repo_valid(f"{repo_entry.url}/{repo_entry.branch}"):
+                print(f"{repo_entry.url}/{repo_entry.branch}")
+                fancy(
+                    "error",
+                    f"File 'packagelist' not found in the '{repo_name}' repository root.",
+                )
+                return (ErrorCodes.NO_HOST_ERROR, None)
+
+            parsed_repo_list.append(repo_entry)
+
+        return (None, parsed_repo_list)
+    except Exception as error:
```

Comment:
Catch specific exceptions.

Reply:
This is just a "catch-all" kind mechanism. Basically this catch should rarely happen.
This applies to all the places where I used `Exception`